internal/proxy/configuration.go

```diff
@@ -0,0 +1,458 @@
+// UPSTREAM_DEFAULT_EMAIL_ADDRESSES
+// UPSTREAM_DEFAULT_GROUPS
+// UPSTREAM_DEFAULT_TIMEOUT
+// UPSTREAM_DEFAULT_TCP_RESET_DEADLINE
+// UPSTREAM_DEFAULT_PROVIDER
+// UPSTREAM_CONFIGS_FILE
+// UPSTREAM_SCHEME
+// UPSTREAM_CLUSTER
+//
+// PROVIDER_TYPE
+// PROVIDER_URL_EXTERNAL
+// PROVIDER_URL_INTERNAL
+// PROVIDER_SCOPE
+
+func DefaultProxyConfig() Configuration {
+	return Configuration{
+		ServerConfig: ServerConfig{
+			Port: 4180,
+			TimeoutConfig: &TimeoutConfig{
+				Write:    30 * time.Second,
+				Read:     30 * time.Second,
+				Shutdown: 30 * time.Second,
+			},
+		},
+		ProviderConfig: ProviderConfig{
+			ProviderType: "sso",
+		},
+		SessionConfig: SessionConfig{
+			CookieConfig: &CookieConfig{
+				Name:     "_sso_proxy",
+				Expire:   168 * time.Hour,
+				Secure:   true,
+				HTTPOnly: true,
+			},
+			TTLConfig: TTLConfig{
+				Lifetime:    720 * time.Hour,
+				Valid:       60 * time.Second,
+				GracePeriod: 3 * time.Hour,
+			},
+		},
+		UpstreamConfigs: UpstreamConfigs{
+			DefaultConfig: DefaultConfig{
+				Timeout:       10 * time.Second,
+				ResetDeadline: 60 * time.Second,
+				ProviderSlug:  "google",
+			},
+			Scheme: "https",
+		},
+		LoggingConfig: LoggingConfig{
+			Enable: true,
+		},
+		MetricsConfig: MetricsConfig{
+			StatsdConfig: StatsdConfig{
+				Port: 8125,
+				Host: "localhost",
+			},
+		},
+	}
+}
+
+type Validator interface {
+	Validate() error
+}
+
+var (
+	_ Validator = Configuration{}
+	_ Validator = ProviderConfig{}
+	_ Validator = SessionConfig{}
+	_ Validator = &CookieConfig{}
+	_ Validator = TTLConfig{}
+	_ Validator = ClientConfig{}
+	_ Validator = ServerConfig{}
+	_ Validator = TimeoutConfig{}
+	_ Validator = MetricsConfig{}
+	_ Validator = StatsdConfig{}
+	_ Validator = LoggingConfig{}
+	_ Validator = UpstreamConfigs{}
+	_ Validator = DefaultConfig{}
+	_ Validator = EmailConfig{}
+	_ Validator = RequestSignerConfig{}
+)
+
+type Configuration struct {
+	ServerConfig        ServerConfig        `mapstructure:"server"`
+	ProviderConfig      ProviderConfig      `mapstructure:"provider"`
+	ClientConfig        ClientConfig        `mapstructure:"client"`
+	SessionConfig       SessionConfig       `mapstructure:"session"`
+	UpstreamConfigs     UpstreamConfigs     `mapstructure:"upstream"`
+	MetricsConfig       MetricsConfig       `mapstructrue:"metrics"`
+	LoggingConfig       LoggingConfig       `mapstructure:"logging"`
+	RequestSignerConfig RequestSignerConfig `mapstructure:"requestsigner"`
+}
+
+func (c Configuration) Validate() error {
+	if err := c.ServerConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid server config: %w", err)
+	}
+
+	if err := c.ProviderConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid provider config: %w", err)
+	}
+
+	if err := c.SessionConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid session config: %w", err)
+	}
+
+	if err := c.ClientConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid client config: %w", err)
```

Comment:
should we add tests for these functions?

Reply:
Individual configuration structs have their `Validate()` method tested in the `TestConfigValidate` function [here](https://github.com/buzzfeed/sso/pull/279/files#diff-b2162c167e6567cc0326346abf552416R101). Though I've also refactored this catch-all `Configuration.Validate()` method slightly to not return after the first error, but continue until all configurations have been tested and return all errors together.

This made it easier to add in a test to validate the functionality within this method (`Configuration.Validate()`) itself.